File: implementations/react/src/patterns/behavioral/strategy/pricing-context.ts
import type { PricingStrategy } from "./strategy-interface";

export class PricingContext {
  private strategy: PricingStrategy;

  constructor(strategy: PricingStrategy) {
    this.strategy = strategy;
  }

  setStrategy(s: PricingStrategy) {
    this.strategy = s;
  }

  getUnitPrice(base: number) {
    return this.strategy.getPrice(base, 1);
  }

  getTotal(base: number, qty: number) {
    return this.strategy.getPrice(base, qty);
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/strategy/pricing_context.dart
import 'package:design_patterns_flutter/patterns/behavioral/strategy/strategy.dart';

class PricingContext {
  PricingStrategy strategy;
  PricingContext(this.strategy);

  void setStrategy(PricingStrategy s) => strategy = s;

  double getUnitPrice(double base) => strategy.getPrice(base, 1);
  double getTotal(double base, int qty) => strategy.getPrice(base, qty);
}
